From: Yury Ilyanok <yury.ilyanok@gmail.com>
Date: Sat, 24 May 2026 12:00:00 +0000
Subject: net: dsa: add Realtek OTTO trailer tag driver

Backport DSA_TAG_PROTO_RTL_OTTO from the OpenWrt realtek target
(kernel 6.18 files).  The RTL9300-series switches use a 4-byte
trailing tag to identify the source/destination port.  The tag driver
implementation is added via the qualcommax files/ overlay.  This tag
protocol enum entry and Makefile hook are required by the out-of-tree
kmod-rtl930x-spi DSA driver used on the Verizon CR1000A.

Signed-off-by: Yury Ilyanok <yury.ilyanok@gmail.com>
---
 include/net/dsa.h | 2 ++
 net/dsa/Makefile  | 1 +
 2 files changed, 3 insertions(+)

--- a/include/net/dsa.h
+++ b/include/net/dsa.h
@@ -57,6 +57,8 @@
 #define DSA_TAG_PROTO_MXL862_VALUE		30
 #define DSA_TAG_PROTO_MXL862_8021Q_VALUE	31
 #define DSA_TAG_PROTO_OOB_VALUE			32
+#define DSA_TAG_PROTO_RTL_OTTO_VALUE		33
+#define DSA_TAG_PROTO_RTL_OTTO_8021Q_VALUE	34

 enum dsa_tag_protocol {
 	DSA_TAG_PROTO_NONE		= DSA_TAG_PROTO_NONE_VALUE,
@@ -91,6 +93,8 @@
 	DSA_TAG_PROTO_MXL862		= DSA_TAG_PROTO_MXL862_VALUE,
 	DSA_TAG_PROTO_MXL862_8021Q	= DSA_TAG_PROTO_MXL862_8021Q_VALUE,
 	DSA_TAG_PROTO_OOB		= DSA_TAG_PROTO_OOB_VALUE,
+	DSA_TAG_PROTO_RTL_OTTO		= DSA_TAG_PROTO_RTL_OTTO_VALUE,
+	DSA_TAG_PROTO_RTL_OTTO_8021Q	= DSA_TAG_PROTO_RTL_OTTO_8021Q_VALUE,
 };

 struct dsa_switch;
--- a/net/dsa/Makefile
+++ b/net/dsa/Makefile
@@ -36,6 +36,8 @@
 obj-$(CONFIG_NET_DSA_TAG_QCA) += tag_qca.o
 obj-$(CONFIG_NET_DSA_TAG_RTL4_A) += tag_rtl4_a.o
 obj-$(CONFIG_NET_DSA_TAG_RTL8_4) += tag_rtl8_4.o
+obj-$(CONFIG_NET_DSA) += tag_rtl_otto.o
+obj-$(CONFIG_NET_DSA) += tag_rtl_otto_8021q.o
 obj-$(CONFIG_NET_DSA_TAG_RZN1_A5PSW) += tag_rzn1_a5psw.o
 obj-$(CONFIG_NET_DSA_TAG_SJA1105) += tag_sja1105.o
 obj-$(CONFIG_NET_DSA_TAG_TRAILER) += tag_trailer.o
